# Hypothesis Testing

The purpose of the test is to tell if there is any significant difference between two data sets.



## Overview

This module covers,

1) One sample and Two sample t-tests

2) ANOVA

3) Type I and Type II errors

4) Chi-Squared Tests

## Question 1 

*A student is trying to decide between two GPUs. He want to use the GPU for his research to run Deep learning algorithms, so the only thing he is concerned with is speed.*

*He picks a Deep Learning algorithm on a large data set and runs it on both GPUs 15 times, timing each run in hours. Results are given in the below lists GPU1 and GPU2.*

In [2]:
from scipy import stats 
import numpy as np

In [3]:
GPU1 = np.array([11,9,10,11,10,12,9,11,12,9,11,12,9,10,9])
GPU2 = np.array([11,13,10,13,12,9,11,12,12,11,12,12,10,11,13])

#Assumption: Both the datasets (GPU1 & GPU 2) are random, independent, parametric & normally distributed

Hint: You can import ttest function from scipy to perform t tests 

**First T test**

*One sample t-test*

Check if the mean of the GPU1 is equal to zero.
- Null Hypothesis is that mean is equal to zero.
- Alternate hypothesis is that it is not equal to zero.

In [6]:
#Null Hypothesis mean 
xbar=np.mean(GPU1)
print("mean of GPU1 = ",xbar)
if xbar == 0:
  print("the mean of the GPU1 is not equal to zero")
else:
  print(xbar)
  print("mean of the GPU1 is not equal to zero")   

mean of GPU1 =  10.333333333333334
10.333333333333334
mean of the GPU1 is not equal to zero


## Question 2

Given,

Null Hypothesis : There is no significant difference between data sets

Alternate Hypothesis : There is a significant difference

*Do two-sample testing and check whether to reject Null Hypothesis or not.*

https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_ind.html

In [11]:
alpha =0.05
Ttest,pvalue=stats.ttest_ind(GPU1,GPU2) #standard independent 2 sample test that assumes equal population variances
#stats.ttest_ind(GPU1,GPU2, equal_var = False) # t-test, which does not assume equal population variance
print(pvalue)
if pvalue<alpha:
    print("Reject Null")
else:
    print("Fail to Reject Null")

0.013794282041452725
Reject Null


As the p value is smaller than the threshold, we reject the null hypothesis. This means both GPUs are not effectively the same speed.

## Question 3

He is trying a third GPU - GPU3.

In [19]:
GPU3 = np.array([9,10,9,11,10,13,12,9,12,12,13,12,13,10,11])

#Assumption: Both the datasets (GPU1 & GPU 3) are random, independent, parametric & normally distributed

*Do two-sample testing and check whether there is significant differene between speeds of two GPUs GPU1 and GPU3.*

#### Answer:

 Step 1: Define null and alternative hypotheses


In testing whether speeds of two GPUs GPU1 and GPU3 are same,the null hypothesis states that mean speed of two GPUs are equal, 𝜇1 = 𝜇3
The alternative hypthesis states that the speed is different for GPU1 and GPU3, μ1 ≠ 𝜇3
H0 : μ1 - 𝜇3 = 0
HA : 𝜇1 - μ3 ≠ 0

Step 2: Decide the significance level.
Here we select α = 0.05 and sample size < 30 and population standard deviation is not known.

Step 3: Identify the test statistic
    Sample sizes for both samples are same.
    We have two paired samples and we do not know the population standard deviation.
    The sample is not a large sample, n < 30. So you use the t distribution and the 𝑡𝑆𝑇𝐴𝑇 test statistic for two sample paired test.

Step 4: Calculate the p - value and test statistic
We use the scipy.stats.ttest_ind to calculate the t-test for the means of TWO INDEPENDENT samples of GPUs given the two sample observations.
This function returns t statistic and  p value.

In [21]:
t_statistic, p_value  =  stats.ttest_ind(GPU1,GPU3)
print('P Value %1.3f' % p_value)    

P Value 0.145


Our p-value is 0.145 which is greater than the common significance value of 0.05. Since it is greater, we fail to reject Null Hypothesis. This means both GPUs are effectively the same speed.

## ANOVA

## Question 4 

If you need to compare more than two data sets at a time, an ANOVA is your best bet. 

*The results from three experiments with overlapping 95% confidence intervals are given below, and we want to confirm that the results for all three experiments are not significantly different.*

But before conducting ANOVA, test equality of variances (using Levene's test) is satisfied or not. If not, then mention that we cannot depend on the result of ANOVA

In [14]:
import numpy as np

e1 = np.array([1.595440,1.419730,0.000000,0.000000])
e2 = np.array([1.433800,2.079700,0.892139,2.384740])
e3 = np.array([0.036930,0.938018,0.995956,1.006970])

#Assumption: All the 3 datasets (e1,e2 & e3) are random, independent, parametric & normally distributed

Perform levene test on the data

The Levene test tests the null hypothesis that all input samples are from populations with equal variances. Levene’s test is an alternative to Bartlett’s test bartlett in the case where there are significant deviations from normality.

source: scipy.org

#### Answer:

In [19]:
import pandas as pd
import scipy.stats as stats

In [20]:
print('Count, Mean and standard deviation of e1: %3d, %3.2f and %3.2f' % (len(e1 ), e1 .mean(),np.std(e1 ,ddof =1)))
print('Count, Mean and standard deviation of e2: %3d, %3.2f and %3.2f' % (len(e2), e2.mean(),np.std(e2,ddof =1)))
print('Count, Mean and standard deviation of e3: %3d, %3.2f and %3.2f' % (len(e3), e3.mean(),np.std(e3,ddof =1)))

Count, Mean and standard deviation of e1:   4, 0.75 and 0.87
Count, Mean and standard deviation of e2:   4, 1.70 and 0.67
Count, Mean and standard deviation of e3:   4, 0.74 and 0.47


In [21]:
array_df = pd.DataFrame()

df1            = pd.DataFrame({'array_NO': '1', 'array_value':e1})
df2            = pd.DataFrame({'array_NO': '2', 'array_value':e2})
df3            = pd.DataFrame({'array_NO': '3', 'array_value':e3})

array_df = array_df.append(df1) 
array_df = array_df.append(df2) 
array_df = array_df.append(df3) 

In [22]:
stats.levene(e1, e2, e3)

LeveneResult(statistic=2.6741725711150446, pvalue=0.12259792666001798)

As p value is higher than .05, we fail to reject null Hypothesis which means that the e1,e2 and e3 pass equality of variance test. So, we can perform ANOVA test.

## Question 5

The one-way ANOVA tests the null hypothesis that two or more groups have the same population mean. The test is applied to samples from two or more groups, possibly with differing sizes.

use stats.f_oneway() module to perform one-way ANOVA test

In [12]:
F, p = stats.f_oneway(e1, e2, e3)
print('F statistic = {:5.3f} and probability p = {:5.3f}'.format(F, p)) 

F statistic = 2.514 and probability p = 0.136


As p value is higher than .05, we fail to reject NULL Hypothesis and we can conclude that e1,e2 and e3 are not significantly different.

we can claim  that the means of the results of all three experiments are not significantly different.

## Question 6

*In one or two sentences explain about **TypeI** and **TypeII** errors.*

#### Answer:

No hypothesis test is 100% certain. Because the test is based on probabilities, there is always a chance of making an incorrect conclusion. When you do a hypothesis test, two types of errors are possible: type I and type II.
Type I error
When the null hypothesis is true and reject it, it makes a type I error. 
Type II error
When the null hypothesis is false and fail to reject it, it makes a type II error.

## Question 7 

You are a manager of a chinese restaurant. You want to determine whether the waiting time to place an order has changed in the past month from its previous population mean value of 4.5 minutes. 
State the null and alternative hypothesis.

#### Answer:


Null Hypothesis : the waiting time to place an order in the past month equals to 4.5 minutes
                    H0 : µN = µo (4.5)
             
Alternate Hypothesis : the waiting time to place an order in the past month not equals to 4.5 minutes
                     HA : µN ≠ µo [µN > µo or µN < µo]
                        

## Chi square test

## Question 8

Let's create a small dataset for dice rolls of four players

In [25]:
import numpy as np

d1 = [5, 8, 3, 8]
d2 = [9, 6, 8, 5]
d3 = [8, 12, 7, 2]
d4 = [4, 16, 7, 3]
d5 = [3, 9, 6, 5]
d6 = [7, 2, 5, 7]

dice = np.array([d1, d2, d3, d4, d5, d6])

run the test using SciPy Stats library

Depending on the test, we are generally looking for a threshold at either 0.05 or 0.01. Our test is significant (i.e. we reject the null hypothesis) if we get a p-value below our threshold.

For our purposes, we’ll use 0.01 as the threshold.

use stats.chi2_contingency() module 

This function computes the chi-square statistic and p-value for the hypothesis test of independence of the observed frequencies in the contingency table

Print the following:

- chi2 stat
- p-value
- degree of freedom
- contingency



In [26]:
from scipy import stats

stats.chi2_contingency(dice)

(23.315671914716496,
 0.07766367301496693,
 15,
 array([[ 5.57419355,  8.20645161,  5.57419355,  4.64516129],
        [ 6.50322581,  9.57419355,  6.50322581,  5.41935484],
        [ 6.73548387,  9.91612903,  6.73548387,  5.61290323],
        [ 6.96774194, 10.25806452,  6.96774194,  5.80645161],
        [ 5.34193548,  7.86451613,  5.34193548,  4.4516129 ],
        [ 4.87741935,  7.18064516,  4.87741935,  4.06451613]]))

In [27]:
chi2_stat, p_val, dof, ex = stats.chi2_contingency(dice)
print("   Chi2 Stat    ")
print(chi2_stat)
print("\n")
print("   Degrees of Freedom    ")
print(dof)
print("  P-Value   ")
print(p_val)
print("\n")
print("   Contingency Table    ")
print(ex)

   Chi2 Stat    
23.315671914716496


   Degrees of Freedom    
15
  P-Value   
0.07766367301496693


   Contingency Table    
[[ 5.57419355  8.20645161  5.57419355  4.64516129]
 [ 6.50322581  9.57419355  6.50322581  5.41935484]
 [ 6.73548387  9.91612903  6.73548387  5.61290323]
 [ 6.96774194 10.25806452  6.96774194  5.80645161]
 [ 5.34193548  7.86451613  5.34193548  4.4516129 ]
 [ 4.87741935  7.18064516  4.87741935  4.06451613]]


Our test is not significant (i.e. we fail to reject the null hypothesis) as we get a p-value equals to 0.07 which isabove our threshold.
Thus we have not met the threshold for statistical significance.

## Question 9

### Z-test

Get zscore on the above dice data using stats.zscore module from scipy. Convert zscore values to p-value and take mean of the array.

In [31]:
z_scores=stats.zscore(dice)
z_scores

array([[-0.46291005, -0.18884739, -1.83711731,  1.44115338],
       [ 1.38873015, -0.64208114,  1.22474487,  0.        ],
       [ 0.9258201 ,  0.7176201 ,  0.61237244, -1.44115338],
       [-0.9258201 ,  1.62408759,  0.61237244, -0.96076892],
       [-1.38873015,  0.03776948,  0.        ,  0.        ],
       [ 0.46291005, -1.54854863, -0.61237244,  0.96076892]])

In [32]:
import scipy
p_values = stats.norm.sf(abs(z_scores))
p_values

array([[0.32171442, 0.42510621, 0.03309629, 0.07477068],
       [0.08245741, 0.26041025, 0.11033568, 0.5       ],
       [0.17726974, 0.23649578, 0.27014569, 0.07477068],
       [0.17726974, 0.05217856, 0.27014569, 0.16833418],
       [0.08245741, 0.48493574, 0.5       , 0.5       ],
       [0.32171442, 0.06074513, 0.27014569, 0.16833418]])

In [33]:
p_values.mean()

0.23428473233691496

## Question 10

A Paired sample t-test compares means from the same group at different times.

The basic two sample t-test is designed for testing differences between independent groups. 
In some cases, you might be interested in testing differences between samples of the same group at different points in time. 
We can conduct a paired t-test using the scipy function stats.ttest_rel(). 

In [36]:
before= stats.norm.rvs(scale=30, loc=100, size=500) ## Creates a normal distribution with a mean value of 100 and std of 30
after = before + stats.norm.rvs(scale=5, loc=-1.25, size=500)

Test whether a weight-loss drug works by checking the weights of the same group patients before and after treatment using above data.

In [38]:
import  scipy.stats  as stats  
stats.ttest_rel(before,after)

Ttest_relResult(statistic=7.006718394972439, pvalue=7.941666437515366e-12)

p value 7.9416e-12 < 0.05 (alpha) so we reject null , means their is significance difference  in the weights of the same group patients before and after treatment.